includes/Pages/UserAuth/Login/LoginCredentialPageBase.php

```diff
@@ -0,0 +1,335 @@
+<?php
+/******************************************************************************
+ * Wikipedia Account Creation Assistance tool                                 *
+ *                                                                            *
+ * All code in this file is released into the public domain by the ACC        *
+ * Development Team. Please see team.json for a list of contributors.         *
+ ******************************************************************************/
+
+namespace Waca\Pages\UserAuth\Login;
+
+use PDO;
+use Waca\DataObjects\User;
+use Waca\Exceptions\ApplicationLogicException;
+use Waca\Exceptions\OAuthException;
+use Waca\Helpers\OAuthUserHelper;
+use Waca\PdoDatabase;
+use Waca\Security\AuthenticationManager;
+use Waca\SessionAlert;
+use Waca\Tasks\InternalPageBase;
+use Waca\WebRequest;
+
+abstract class LoginCredentialPageBase extends InternalPageBase
+{
+    /** @var User */
+    protected $partialUser = null;
+    protected $nextPageMap = array(
+        'yubikeyotp' => 'otp',
+        'totp'       => 'otp',
+        'scratch'    => 'otp',
+        'u2f'        => 'u2f',
+    );
+    protected $names = array(
+        'yubikeyotp' => 'Yubikey OTP',
+        'totp'       => 'TOTP (phone code generator)',
+        'scratch'    => 'scratch token',
+        'u2f'        => 'U2F security token',
+    );
+
+    /**
+     * Main function for this page, when no specific actions are called.
+     * @return void
+     */
+    protected function main()
+    {
+        if (!$this->enforceHttps()) {
+            return;
+        }
+
+        if (WebRequest::wasPosted()) {
+            $this->validateCSRFToken();
+
+            $database = $this->getDatabase();
+            try {
+                list($partialId, $partialStage) = WebRequest::getAuthPartialLogin();
+
+                if ($partialStage === null) {
+                    $partialStage = 1;
+                }
+
+                if ($partialId === null) {
+                    $username = WebRequest::postString('username');
+
+                    if ($username === null || trim($username) === '') {
+                        throw new ApplicationLogicException('No username specified.');
+                    }
+
+                    $user = User::getByUsername($username, $database);
+                }
+                else {
+                    $user = User::getById($partialId, $database);
+                }
+
+                if ($user === false) {
+                    throw new ApplicationLogicException("Authentication failed");
```

Comment:
This could really use a more descriptive message. I assume the only use case is when the user is not found?